src/perception/zed_wrapper/zed_wrapper.cpp

```diff
@@ -0,0 +1,258 @@
+            mPnh.param("image_height", imageHeight, 720);
+            mPnh.param("depth_confidence", mDepthConfidence, 70);
+            mPnh.param("texture_confidence", mTextureConfidence, 100);
+            mPnh.param("direct_tag_detection", mDirectTagDetection, false);
+            std::string svoFile{};
+            mPnh.param("svo_file", svoFile, {});
+            mPnh.param("ues_builtin_visual_odom", mUseBuiltinPosTracking, false);
+
+            if (imageWidth < 0 || imageHeight < 0) {
+                throw std::invalid_argument("Invalid image dimensions");
+            }
+            if (mGrabTargetFps < 0) {
+                throw std::invalid_argument("Invalid grab target framerate");
+            }
+
+            sl::InitParameters initParameters;
+            if (!svoFile.empty()) {
+                initParameters.input.setFromSVOFile(svoFile.c_str());
+            }
+            initParameters.camera_resolution = static_cast<sl::RESOLUTION>(resolution);
+            initParameters.depth_mode = static_cast<sl::DEPTH_MODE>(depthMode);
+            initParameters.coordinate_units = sl::UNIT::METER;
+            initParameters.sdk_verbose = true; // Log useful information
+            initParameters.camera_fps = mGrabTargetFps;
+            initParameters.coordinate_system = sl::COORDINATE_SYSTEM::RIGHT_HANDED_Z_UP_X_FWD; // Match ROS
+
+            if (mZed.open(initParameters) != sl::ERROR_CODE::SUCCESS) {
+                throw std::runtime_error("ZED failed to open");
+            }
+
+            sl::PositionalTrackingParameters positionalTrackingParameters;
+            mZed.enablePositionalTracking(positionalTrackingParameters);
+
+            mGrabThread = std::thread(&ZedNodelet::grabUpdate, this);
+            mPcThread = std::thread(&ZedNodelet::pointCloudUpdate, this);
+
+        } catch (std::exception const& e) {
+            NODELET_FATAL("Exception while starting: %s", e.what());
+            ros::shutdown();
+        }
+    }
+
+    /**
+     * Only relevant if direct tag detection is true.
+     * This handles loading a tag detection nodelet and passing it point clouds from the grab thread.
+     */
+    void ZedNodelet::pointCloudUpdate() {
+        try {
+            NODELET_INFO("Starting point cloud thread");
+
+            if (mDirectTagDetection) {
+                // TODO: ugly, this prevents OpenCV fast alloc from crashing
+                std::this_thread::sleep_for(100ms);
+
+                mTagDetectorNode = boost::make_shared<TagDetectorNodelet>();
+                mTagDetectorNode->init("tag_detector", getRemappingArgs(), getMyArgv());
+
+                // TODO: figure out why removing this causes a segfault
+                cv::aruco::getPredefinedDictionary(cv::aruco::DICT_4X4_50);
+            }
+
+            while (ros::ok()) {
+                mPcThreadProfiler.reset();
+
+                {
+                    std::unique_lock lock{mGrabMutex};
+                    mGrabDone.wait(lock);
```

Comment:
cv wait not in while loop?

Reply:
I could add a variable that is checked but the ZED library is already thread-safe so it's not a big deal if it wakes up every once in a while

Reply:
oh. Okay as long as you're 100% sure that calling retrieveMeasure/Image is thread safe with grab thats fine.